Add integration test of static policy plugin

bakerbrandond opened the pull request:
- Asserts that expected site policy is enforced on non-GEOPM jobs
- Fixes #874

Change-Id: I32b4197323e2814e20ba122ab4f32b6ca2eef604
Signed-off-by: Brandon Baker <brandon.baker@intel.com>

Commit:
Add integration test of static policy plugin

- Asserts that expected site policy is enforced on non-GEOPM jobs
- Fixes #874

Change-Id: I32b4197323e2814e20ba122ab4f32b6ca2eef604
Signed-off-by: Brandon Baker <brandon.baker@intel.com>

bgeltz commented on test_integration/test_plugin_static_policy.py:
```diff
@@ -0,0 +1,115 @@
+
+# todo this function results in test launcher
+# invocation, and is called 3 times (1 by class and
+# in 2 decorator tests).  may be useful to add once logic
```
Is there an issue open for this?

bgeltz replied:
Is there an issue open for this open already?

bgeltz commented on test_integration/test_plugin_static_policy.py:
```diff
@@ -0,0 +1,115 @@
+        cls._geopmadminagentpolicy = geopmadminagentpolicy()
+
+    @unittest.skipUnless(geopmadminagent() in ['frequency_map', 'energy_efficient'],
+                         'Requires environment defaut/override to be configured to cap frequency.')
```
Spelling: defau*L*t/override

bgeltz commented on test_integration/test_plugin_static_policy.py:
```diff
@@ -0,0 +1,115 @@
+                         'Requires environment defaut/override to be configured to cap power.')
+    def test_power_cap_enforced(self):
+        test_power = self._geopmadminagentpolicy['POWER_PACKAGE_LIMIT_TOTAL']
+        # todo need test_launcher to execute topo on compute node
```
Issue for this too please.

bakerbrandond replied:
added a decorator to the test class to require the test be ran in an allocation to resolve this (same as the static (non plugin) policy integration test)

bgeltz commented on test_integration/test_plugin_static_policy.py:
```diff
@@ -0,0 +1,115 @@
+        test_power = self._geopmadminagentpolicy['POWER_PACKAGE_LIMIT_TOTAL']
+        # todo need test_launcher to execute topo on compute node
+        num_pkg = geopmpy.topo.num_domain('package')
+        current_power = num_pkg * geopm_test_launcher.geopmread("MSR::PKG_POWER_LIMIT:PL1_POWER_LIMIT board 0")
```
It isn't safe to assume that the PL1 limit is the same on both sockets if multisocket.

I was going to suggest you use num_pkg to loop over the packages, and use geopmread to read the limit on the individual sockets.  But I see you're reading the board limit...

I think this should fail:  reading the board limit should read each package and sum them I think.  There's a bug somewhere here, I think it's in how geopmread sums (or not) the values for the board domain if this logic works.

bakerbrandond replied:
inherited from here: https://github.com/geopm/geopm/blob/dev/test_integration/test_static_policy.py#L173-L180
shall we file a bug?

bakerbrandond replied:
There is a subtlety here that we both missed.  The policy name is: POWER_PACKAGE_LIMIT_TOTAL, as in total for all of the packages on the board.  That's why sampling board works but yes this implies they packages split the allotted power equally.  Is a non uniform distribution of power across packages expected to appear in the wild?

bgeltz reviewed (changes requested):
A few nits, and a (I think) a bonafide bug.

Commit:
Add static member to Environment base class for environment map access

- Refactor EnvironmentImp::parse_environment_file such that it
  is statically accessible.

Change-Id: I09903e7254942f899946a56679e0ea5c8f60558e
Signed-off-by: Brandon Baker <brandon.baker@intel.com>

cmcantalupo commented on src/Environment.hpp:
```diff
@@ -107,13 +108,16 @@ namespace geopm
             bool do_ompt(void) const override;
             std::string default_config_path(void) const override;
             std::string override_config_path(void) const override;
+            static void parse_environment_file(const std::string &settings_path,
```
This static method could be private right?

bakerbrandond replied:
Nope the Environment (base class) needs access to this static public member of the EnvironmentImp class line 86 in green of this patch

cmcantalupo reviewed (commented):
That should work

bgeltz reviewed (commented):
The integration test looks good now.  Thanks for fixing that up.  I think that bit is g2g.

I assume you and Chris are hashing out the rest, so I didn't approve.
